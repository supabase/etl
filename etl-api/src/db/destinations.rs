use etl_config::SerializableSecretString;
use etl_config::shared::DestinationConfig;
use secrecy::ExposeSecret;
use serde::{Deserialize, Serialize};
use sqlx::PgExecutor;
use std::fmt::Debug;
use thiserror::Error;

use crate::db::serde::{
    DbDeserializationError, DbSerializationError, decrypt_and_deserialize_from_value,
    encrypt_and_serialize,
};
use crate::encryption::{
    Decrypt, DecryptionError, Encrypt, EncryptedValue, EncryptionError, EncryptionKey,
    decrypt_text, encrypt_text,
};

impl Encrypt<EncryptedDestinationConfig> for DestinationConfig {
    fn encrypt(
        self,
        encryption_key: &EncryptionKey,
    ) -> Result<EncryptedDestinationConfig, EncryptionError> {
        match self {
            Self::Memory => Ok(EncryptedDestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let encrypted_service_account_key = encrypt_text(
                    service_account_key.expose_secret().to_owned(),
                    encryption_key,
                )?;

                Ok(EncryptedDestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key: encrypted_service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EncryptedDestinationConfig {
    Memory,
    BigQuery {
        project_id: String,
        dataset_id: String,
        service_account_key: EncryptedValue,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_staleness_mins: Option<u16>,
        #[serde(skip_serializing_if = "Option::is_none")]
        max_concurrent_streams: Option<usize>,
    },
}

impl Decrypt<DestinationConfig> for EncryptedDestinationConfig {
    fn decrypt(self, encryption_key: &EncryptionKey) -> Result<DestinationConfig, DecryptionError> {
        match self {
            Self::Memory => Ok(DestinationConfig::Memory),
            Self::BigQuery {
                project_id,
                dataset_id,
                service_account_key: encrypted_service_account_key,
                max_staleness_mins,
                max_concurrent_streams,
            } => {
                let service_account_key = SerializableSecretString::from(decrypt_text(
                    encrypted_service_account_key,
                    encryption_key,
                )?);

                Ok(DestinationConfig::BigQuery {
                    project_id,
                    dataset_id,
                    service_account_key,
                    max_staleness_mins,
                    max_concurrent_streams,
                })
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DestinationsDbError {
    #[error("Error while interacting with Postgres for destinations: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing destination config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing destination config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),
}

pub struct Destination {
    pub id: i64,
    pub tenant_id: String,
    pub name: String,
    pub config: DestinationConfig,
}

pub async fn create_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    name: &str,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<i64, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let config = encrypt_and_serialize(config, encryption_key)?;

    let record = sqlx::query!(
        r#"
        insert into app.destinations (tenant_id, name, config)
        values ($1, $2, $3)
        returning id
        "#,
        tenant_id,
        name,
        config
    )
    .fetch_one(executor)
    .await?;

    Ok(record.id)
}

pub async fn read_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
    encryption_key: &EncryptionKey,
) -> Result<Option<Destination>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1 and id = $2
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_optional(executor)
    .await?;

    let destination = match record {
        Some(record) => {
            let config = decrypt_and_deserialize_from_value::<
                EncryptedDestinationConfig,
                DestinationConfig,
            >(record.config, encryption_key)?;

            let destination = Destination {
                id: record.id,
                tenant_id: record.tenant_id,
                name: record.name,
                config,
            };

            Some(destination)
        }
        None => None,
    };

    Ok(destination)
}

pub async fn update_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    name: &str,
    destination_id: i64,
    config: DestinationConfig,
    encryption_key: &EncryptionKey,
) -> Result<Option<i64>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let config = encrypt_and_serialize(config, encryption_key)?;

    let record = sqlx::query!(
        r#"
        update app.destinations
        set config = $1, name = $2
        where tenant_id = $3 and id = $4
        returning id
        "#,
        config,
        name,
        tenant_id,
        destination_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_destination<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
) -> Result<Option<i64>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.destinations
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        destination_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_destinations<'c, E>(
    executor: E,
    tenant_id: &str,
    encryption_key: &EncryptionKey,
) -> Result<Vec<Destination>, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let records = sqlx::query!(
        r#"
        select id, tenant_id, name, config
        from app.destinations
        where tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(executor)
    .await?;

    let mut destinations = Vec::with_capacity(records.len());
    for record in records {
        let config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(record.config.clone(), encryption_key)?;

        let destination = Destination {
            id: record.id,
            tenant_id: record.tenant_id,
            name: record.name,
            config,
        };
        destinations.push(destination);
    }

    Ok(destinations)
}

pub async fn destination_exists<'c, E>(
    executor: E,
    tenant_id: &str,
    destination_id: i64,
) -> Result<bool, DestinationsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select exists (select id
        from app.destinations
        where tenant_id = $1 and id = $2) as "exists!"
        "#,
        tenant_id,
        destination_id,
    )
    .fetch_one(executor)
    .await?;

    Ok(record.exists)
}

#[cfg(test)]
mod tests {
    use aws_lc_rs::aead::RandomizedNonceKey;
    use etl_config::SerializableSecretString;
    use etl_config::shared::DestinationConfig;

    use crate::db::destinations::EncryptedDestinationConfig;
    use crate::db::serde::{decrypt_and_deserialize_from_value, encrypt_and_serialize};
    use crate::encryption::EncryptionKey;

    #[test]
    pub fn destination_config_json_deserialization() {
        let config = r#"{
            "big_query": {
                "project_id": "project-id",
                "dataset_id": "dataset-id",
                "service_account_key": "service-account-key",
                "max_staleness_mins": 42
            }
        }"#;

        let deserialized = serde_json::from_str::<DestinationConfig>(config);
        insta::assert_debug_snapshot!(deserialized.unwrap());
    }

    #[test]
    pub fn destination_config_json_serialization() {
        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: SerializableSecretString::from("service-account-key".to_string()),
            max_staleness_mins: Some(42),
            max_concurrent_streams: None,
        };

        insta::assert_json_snapshot!(config);
    }

    #[test]
    pub fn destination_config_json_encryption() {
        let key_bytes = [42u8; 32];
        let key = RandomizedNonceKey::new(&aws_lc_rs::aead::AES_256_GCM, &key_bytes).unwrap();
        let encryption_key = EncryptionKey { id: 1, key };

        let config = DestinationConfig::BigQuery {
            project_id: "project-id".to_string(),
            dataset_id: "dataset-id".to_string(),
            service_account_key: SerializableSecretString::from("supersecretkey".to_string()),
            max_staleness_mins: Some(99),
            max_concurrent_streams: None,
        };

        let config_in_db = encrypt_and_serialize::<DestinationConfig, EncryptedDestinationConfig>(
            config.clone(),
            &encryption_key,
        )
        .unwrap();
        insta::assert_json_snapshot!(config_in_db, {
            ".big_query.service_account_key" => "[key]"
        });

        let deserialized_config = decrypt_and_deserialize_from_value::<
            EncryptedDestinationConfig,
            DestinationConfig,
        >(config_in_db, &encryption_key)
        .unwrap();
        insta::assert_debug_snapshot!(deserialized_config);
    }
}
