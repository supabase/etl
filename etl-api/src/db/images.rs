use sqlx::{PgExecutor, PgPool};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ImagesDbError {
    #[error("Error while interacting with Postgres for images: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Cannot delete the default image")]
    CannotDeleteDefault,
}

pub struct Image {
    pub id: i64,
    pub name: String,
    pub is_default: bool,
}

pub async fn create_image(
    pool: &PgPool,
    name: &str,
    is_default: bool,
) -> Result<i64, ImagesDbError> {
    let mut txn = pool.begin().await?;

    // If this is to be the new default image, first unset any existing default
    if is_default {
        sqlx::query!(
            r#"
            update app.images
            set is_default = false
            where is_default = true
            "#
        )
        .execute(&mut *txn)
        .await?;
    }

    let record = sqlx::query!(
        r#"
        insert into app.images (name, is_default)
        values ($1, $2)
        returning id
        "#,
        name,
        is_default
    )
    .fetch_one(&mut *txn)
    .await?;

    txn.commit().await?;

    Ok(record.id)
}

pub async fn read_default_image<'c, E>(executor: E) -> Result<Option<Image>, ImagesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        where is_default = true
        "#,
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}

pub async fn read_image<'c, E>(executor: E, image_id: i64) -> Result<Option<Image>, ImagesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        where id = $1
        "#,
        image_id,
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}

pub async fn update_image(
    pool: &PgPool,
    image_id: i64,
    name: &str,
    is_default: bool,
) -> Result<Option<i64>, ImagesDbError> {
    let mut txn = pool.begin().await?;

    // If this is to be the new default image, first unset any existing default
    if is_default {
        sqlx::query!(
            r#"
            update app.images
            set is_default = false
            where is_default = true and id != $1
            "#,
            image_id
        )
        .execute(&mut *txn)
        .await?;
    }

    let record = sqlx::query!(
        r#"
        update app.images
        set name = $1, is_default = $2
        where id = $3
        returning id
        "#,
        name,
        is_default,
        image_id
    )
    .fetch_optional(&mut *txn)
    .await?;

    txn.commit().await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_image(pool: &PgPool, image_id: i64) -> Result<Option<i64>, ImagesDbError> {
    let mut txn = pool.begin().await?;

    // First check if the image exists and if it's a default
    let image = sqlx::query!(
        r#"
        select id, is_default
        from app.images
        where id = $1
        "#,
        image_id
    )
    .fetch_optional(&mut *txn)
    .await?;

    let image = match image {
        Some(img) => img,
        None => {
            // Image doesn't exist, return None
            return Ok(None);
        }
    };

    // Prevent deletion if it's a default image
    if image.is_default {
        return Err(ImagesDbError::CannotDeleteDefault);
    }

    // Proceed with deletion
    let record = sqlx::query!(
        r#"
        delete from app.images
        where id = $1
        returning id
        "#,
        image_id
    )
    .fetch_optional(&mut *txn)
    .await?;

    txn.commit().await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_images<'c, E>(executor: E) -> Result<Vec<Image>, ImagesDbError>
where
    E: PgExecutor<'c>,
{
    let mut record = sqlx::query!(
        r#"
        select id, name, is_default
        from app.images
        "#,
    )
    .fetch_all(executor)
    .await?;

    Ok(record
        .drain(..)
        .map(|r| Image {
            id: r.id,
            name: r.name,
            is_default: r.is_default,
        })
        .collect())
}

pub async fn read_image_by_replicator_id<'c, E>(
    executor: E,
    replicator_id: i64,
) -> Result<Option<Image>, ImagesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select i.id, i.name, i.is_default
        from app.images i
        join app.replicators r on i.id = r.image_id
        where r.id = $1
        "#,
        replicator_id,
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| Image {
        id: r.id,
        name: r.name,
        is_default: r.is_default,
    }))
}
