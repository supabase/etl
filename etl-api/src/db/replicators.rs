use sqlx::PgExecutor;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum ReplicatorsDbError {
    #[error("Error while interacting with PostgreSQL for replicators: {0}")]
    Database(#[from] sqlx::Error),
}

pub struct Replicator {
    pub id: i64,
    pub tenant_id: String,
    pub image_id: i64,
}

pub async fn create_replicator<'c, E>(
    executor: E,
    tenant_id: &str,
    image_id: i64,
) -> Result<i64, ReplicatorsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        insert into app.replicators (tenant_id, image_id)
        values ($1, $2)
        returning id
        "#,
        tenant_id,
        image_id
    )
    .fetch_one(executor)
    .await?;

    Ok(record.id)
}

pub async fn read_replicator_by_pipeline_id<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<Replicator>, ReplicatorsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select r.id, r.tenant_id, r.image_id
        from app.replicators r
        join app.pipelines p on r.id = p.replicator_id
        where r.tenant_id = $1 and p.tenant_id = $1 and p.id = $2
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| Replicator {
        id: r.id,
        tenant_id: r.tenant_id,
        image_id: r.image_id,
    }))
}

pub async fn read_replicators<'c, E>(
    executor: E,
    tenant_id: &str,
) -> Result<Vec<Replicator>, ReplicatorsDbError>
where
    E: PgExecutor<'c>,
{
    let mut records = sqlx::query!(
        r#"
        select r.id, r.tenant_id, r.image_id
        from app.replicators r
        join app.pipelines p on r.id = p.replicator_id
        where r.tenant_id = $1 and p.tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(executor)
    .await?;

    Ok(records
        .drain(..)
        .map(|r| Replicator {
            id: r.id,
            tenant_id: r.tenant_id,
            image_id: r.image_id,
        })
        .collect())
}

pub async fn update_replicator_image<'c, E>(
    executor: E,
    tenant_id: &str,
    replicator_id: i64,
    image_id: i64,
) -> Result<Option<i64>, ReplicatorsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        update app.replicators
        set image_id = $1
        where id = $2 and tenant_id = $3
        returning id
        "#,
        image_id,
        replicator_id,
        tenant_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_replicator<'c, E>(
    executor: E,
    tenant_id: &str,
    replicator_id: i64,
) -> Result<Option<i64>, ReplicatorsDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.replicators
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        replicator_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}
