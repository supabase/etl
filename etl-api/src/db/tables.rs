use serde::{Deserialize, Serialize};
use sqlx::{Executor, PgPool, Row};
use thiserror::Error;
use utoipa::ToSchema;

#[derive(Debug, Error)]
pub enum TablesDbError {
    #[error("Error while interacting with PostgreSQL for tables: {0}")]
    Database(#[from] sqlx::Error),
}

#[derive(Debug, Serialize, Deserialize, ToSchema)]
pub struct Table {
    pub schema: String,
    pub name: String,
}

pub async fn get_tables(pool: &PgPool) -> Result<Vec<Table>, TablesDbError> {
    let query = r#"
        select
           	n.nspname as schema,
           	c.relname as name
        from pg_catalog.pg_class c
           	left join pg_catalog.pg_namespace n on n.oid = c.relnamespace
            left join pg_catalog.pg_am am on am.oid = c.relam
        where
           	c.relkind = 'r'
           	and n.nspname <> 'pg_catalog'
            and n.nspname !~ '^pg_toast'
            and n.nspname <> 'information_schema'
           	and pg_catalog.pg_table_is_visible(c.oid)
        order by schema, name;
        "#;

    let tables = pool
        .fetch_all(query)
        .await?
        .iter()
        .map(|r| Table {
            schema: r.get("schema"),
            name: r.get("name"),
        })
        .collect();

    Ok(tables)
}
