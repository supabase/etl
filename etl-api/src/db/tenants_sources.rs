use sqlx::PgTransaction;
use std::ops::DerefMut;
use thiserror::Error;

use crate::configs::encryption::EncryptionKey;
use crate::configs::serde::DbSerializationError;
use crate::configs::source::FullApiSourceConfig;
use crate::db::sources::{SourcesDbError, create_source};
use crate::db::tenants::{TenantsDbError, create_tenant};

#[derive(Debug, Error)]
pub enum TenantSourceDbError {
    #[error("Error while interacting with Postgres for tenants and/or sources: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing tenant or source config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error(transparent)]
    Sources(#[from] SourcesDbError),

    #[error(transparent)]
    Tenants(#[from] TenantsDbError),
}

pub async fn create_tenant_and_source(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    tenant_name: &str,
    source_name: &str,
    source_config: FullApiSourceConfig,
    encryption_key: &EncryptionKey,
) -> Result<(String, i64), TenantSourceDbError> {
    let tenant_id = create_tenant(txn.deref_mut(), tenant_id, tenant_name).await?;
    let source_id = create_source(
        txn.deref_mut(),
        &tenant_id,
        source_name,
        source_config,
        encryption_key,
    )
    .await?;

    Ok((tenant_id, source_id))
}
