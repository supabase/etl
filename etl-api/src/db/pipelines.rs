use etl_config::shared::BatchConfig;
use serde::{Deserialize, Serialize};
use sqlx::{PgExecutor, PgTransaction};
use std::ops::DerefMut;
use thiserror::Error;
use utoipa::ToSchema;

use crate::db::replicators::{ReplicatorsDbError, create_replicator};
use crate::db::serde::{
    DbDeserializationError, DbSerializationError, deserialize_from_value, serialize,
};

/// Pipeline configuration used during replication. This struct's fields
/// should be kept in sync with [`OptionalPipelineConfig`]. If a new optional
/// field is added, it should also be included in the [`OptionalPipelineConfig::merge`]
/// implementation.
///
/// A separate struct was created because `publication_name` is not optional and
/// when updating config we do not want the user to pass publication
/// name.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct PipelineConfig {
    #[schema(example = "my_publication")]
    pub publication_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch: Option<BatchConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub table_error_retry_delay_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[schema(example = 4)]
    pub max_table_sync_workers: Option<u16>,
}

/// Has the same fields as [`PipelineConfig`] except from
/// the required fields. These two structs should be kept
/// in sync.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct OptionalPipelineConfig {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub batch: Option<BatchConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub table_error_retry_delay_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_table_sync_workers: Option<u16>,
}

impl PipelineConfig {
    /// Merges an [`OptionalPipelineConfig`] into this one by overwriting the optional
    /// fields in self if they are set in the other config. There is currently no
    /// way to unset fields in self, but should be good enough for now.
    fn merge(&mut self, other: OptionalPipelineConfig) {
        if let Some(batch) = other.batch {
            self.batch = Some(batch);
        }

        if let Some(table_error_retry_delay_ms) = other.table_error_retry_delay_ms {
            self.table_error_retry_delay_ms = Some(table_error_retry_delay_ms);
        }

        if let Some(max_table_sync_workers) = other.max_table_sync_workers {
            self.max_table_sync_workers = Some(max_table_sync_workers);
        }
    }
}

pub struct Pipeline {
    pub id: i64,
    pub tenant_id: String,
    pub source_id: i64,
    pub source_name: String,
    pub destination_id: i64,
    pub destination_name: String,
    pub replicator_id: i64,
    pub config: PipelineConfig,
}

#[derive(Debug, Error)]
pub enum PipelinesDbError {
    #[error("Error while interacting with PostgreSQL for pipelines: {0}")]
    Database(#[from] sqlx::Error),

    #[error("Error while serializing pipeline config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing pipeline config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),

    #[error(transparent)]
    ReplicatorsDb(#[from] ReplicatorsDbError),
}

pub async fn create_pipeline(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    source_id: i64,
    destination_id: i64,
    image_id: i64,
    config: PipelineConfig,
) -> Result<i64, PipelinesDbError> {
    let config = serialize(&config)?;

    let replicator_id = create_replicator(txn.deref_mut(), tenant_id, image_id).await?;
    let record = sqlx::query!(
        r#"
        insert into app.pipelines (tenant_id, source_id, destination_id, replicator_id, config)
        values ($1, $2, $3, $4, $5)
        returning id
        "#,
        tenant_id,
        source_id,
        destination_id,
        replicator_id,
        config
    )
    .fetch_one(txn.deref_mut())
    .await?;

    Ok(record.id)
}

pub async fn read_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<Pipeline>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        select p.id,
            p.tenant_id,
            source_id,
            s.name as source_name,
            destination_id,
            d.name as destination_name,
            replicator_id,
            p.config
        from app.pipelines p
        join app.sources s on p.source_id = s.id
        join app.destinations d on p.destination_id = d.id
        where p.tenant_id = $1 and p.id = $2
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(executor)
    .await?;

    let pipeline = match record {
        Some(record) => {
            let config = deserialize_from_value::<PipelineConfig>(record.config)?;

            let pipeline = Pipeline {
                id: record.id,
                tenant_id: record.tenant_id,
                source_id: record.source_id,
                source_name: record.source_name,
                destination_id: record.destination_id,
                destination_name: record.destination_name,
                replicator_id: record.replicator_id,
                config,
            };

            Some(pipeline)
        }
        None => None,
    };

    Ok(pipeline)
}

pub async fn update_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
    source_id: i64,
    destination_id: i64,
    config: &PipelineConfig,
) -> Result<Option<i64>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let pipeline_config = serialize(config)?;

    let record = sqlx::query!(
        r#"
        update app.pipelines
        set source_id = $1, destination_id = $2, config = $3
        where tenant_id = $4 and id = $5
        returning id
        "#,
        source_id,
        destination_id,
        pipeline_config,
        tenant_id,
        pipeline_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn delete_pipeline<'c, E>(
    executor: E,
    tenant_id: &str,
    pipeline_id: i64,
) -> Result<Option<i64>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let record = sqlx::query!(
        r#"
        delete from app.pipelines
        where tenant_id = $1 and id = $2
        returning id
        "#,
        tenant_id,
        pipeline_id
    )
    .fetch_optional(executor)
    .await?;

    Ok(record.map(|r| r.id))
}

pub async fn read_all_pipelines<'c, E>(
    executor: E,
    tenant_id: &str,
) -> Result<Vec<Pipeline>, PipelinesDbError>
where
    E: PgExecutor<'c>,
{
    let records = sqlx::query!(
        r#"
        select p.id,
            p.tenant_id,
            source_id,
            s.name as source_name,
            destination_id,
            d.name as destination_name,
            replicator_id,
            p.config
        from app.pipelines p
        join app.sources s on p.source_id = s.id
        join app.destinations d on p.destination_id = d.id
        where p.tenant_id = $1
        "#,
        tenant_id,
    )
    .fetch_all(executor)
    .await?;

    let mut pipelines = Vec::with_capacity(records.len());
    for record in records {
        let config = deserialize_from_value::<PipelineConfig>(record.config.clone())
            .expect("failed to deserialize pipeline config");

        pipelines.push(Pipeline {
            id: record.id,
            tenant_id: record.tenant_id,
            source_id: record.source_id,
            source_name: record.source_name,
            destination_id: record.destination_id,
            destination_name: record.destination_name,
            replicator_id: record.replicator_id,
            config,
        });
    }

    Ok(pipelines)
}

pub async fn update_pipeline_config(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    pipeline_id: i64,
    request_config: OptionalPipelineConfig,
) -> Result<Option<PipelineConfig>, PipelinesDbError> {
    // We use `select ... for update` to lock the pipeline row being updated
    // to avoid concurrent requests clobbering data from each other
    let record = sqlx::query!(
        r#"
        select p.id,
            p.config
        from app.pipelines p
        where p.tenant_id = $1 and p.id = $2
        for update
        "#,
        tenant_id,
        pipeline_id,
    )
    .fetch_optional(txn.deref_mut())
    .await?;

    match record {
        Some(record) => {
            let mut config_in_db = deserialize_from_value::<PipelineConfig>(record.config)?;
            config_in_db.merge(request_config);
            let updated_config = serialize(config_in_db)?;

            let record = sqlx::query!(
                r#"
                update app.pipelines
                set config = $1
                where tenant_id = $2 and id = $3
                returning config
                "#,
                updated_config,
                tenant_id,
                pipeline_id
            )
            .fetch_optional(txn.deref_mut())
            .await?;

            match record {
                Some(record) => {
                    let config = deserialize_from_value::<PipelineConfig>(record.config)?;
                    Ok(Some(config))
                }
                None => Ok(None),
            }
        }
        None => Ok(None),
    }
}

/// Helper function to check if an sqlx error is a duplicate pipeline constraint violation
pub fn is_duplicate_pipeline_error(err: &sqlx::Error) -> bool {
    match err {
        sqlx::Error::Database(db_err) => {
            // 23505 is PostgreSQL's unique constraint violation code
            // Check for our unique constraint name defined
            // in the migrations/20250605064229_add_unique_constraint_pipelines_source_destination.sql file
            db_err.code().as_deref() == Some("23505")
                && db_err.constraint() == Some("pipelines_tenant_source_destination_unique")
        }
        _ => false,
    }
}
