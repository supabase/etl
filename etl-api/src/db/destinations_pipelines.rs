use config::shared::DestinationConfig;
use sqlx::PgTransaction;
use std::ops::DerefMut;
use thiserror::Error;

use crate::db::destinations::{DestinationsDbError, create_destination, update_destination};
use crate::db::pipelines::{PipelineConfig, PipelinesDbError, create_pipeline, update_pipeline};
use crate::db::serde::{DbDeserializationError, DbSerializationError};
use crate::encryption::EncryptionKey;

#[derive(Debug, Error)]
pub enum DestinationPipelinesDbError {
    #[error("Error while interacting with PostgreSQL for destination and/or pipelines: {0}")]
    Database(#[from] sqlx::Error),

    #[error("The destination with id {0} was not found")]
    DestinationNotFound(i64),

    #[error("The pipeline with id {0} was not found")]
    PipelineNotFound(i64),

    #[error(transparent)]
    PipelinesDb(#[from] PipelinesDbError),

    #[error(transparent)]
    DestinationsDb(#[from] DestinationsDbError),

    #[error("Error while serializing destination or pipeline config: {0}")]
    DbSerialization(#[from] DbSerializationError),

    #[error("Error while deserializing destination or pipeline config: {0}")]
    DbDeserialization(#[from] DbDeserializationError),
}

#[expect(clippy::too_many_arguments)]
pub async fn create_destination_and_pipeline(
    txn: &mut PgTransaction<'_>,
    tenant_id: &str,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    image_id: i64,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(i64, i64), DestinationPipelinesDbError> {
    let destination_id = create_destination(
        txn.deref_mut(),
        tenant_id,
        destination_name,
        destination_config,
        encryption_key,
    )
    .await?;

    let pipeline_id = create_pipeline(
        txn,
        tenant_id,
        source_id,
        destination_id,
        image_id,
        pipeline_config,
    )
    .await?;

    Ok((destination_id, pipeline_id))
}

#[expect(clippy::too_many_arguments)]
pub async fn update_destination_and_pipeline(
    mut txn: PgTransaction<'_>,
    tenant_id: &str,
    destination_id: i64,
    pipeline_id: i64,
    source_id: i64,
    destination_name: &str,
    destination_config: DestinationConfig,
    pipeline_config: PipelineConfig,
    encryption_key: &EncryptionKey,
) -> Result<(), DestinationPipelinesDbError> {
    let destination_id_res = update_destination(
        txn.deref_mut(),
        tenant_id,
        destination_name,
        destination_id,
        destination_config,
        encryption_key,
    )
    .await?;

    if destination_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelinesDbError::DestinationNotFound(
            destination_id,
        ));
    };

    let pipeline_id_res = update_pipeline(
        txn.deref_mut(),
        tenant_id,
        pipeline_id,
        source_id,
        destination_id,
        &pipeline_config,
    )
    .await?;

    if pipeline_id_res.is_none() {
        txn.rollback().await?;
        return Err(DestinationPipelinesDbError::PipelineNotFound(pipeline_id));
    };
    txn.commit().await?;

    Ok(())
}
