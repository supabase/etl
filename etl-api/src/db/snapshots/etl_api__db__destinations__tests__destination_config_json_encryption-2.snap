---
source: etl-api/src/db/destinations.rs
expression: deserialized_config
---
BigQuery {
    project_id: "project-id",
    dataset_id: "dataset-id",
    service_account_key: SerializableSecretString(
        SecretBox<str>([REDACTED]),
    ),
    max_staleness_mins: Some(
        99,
    ),
}
